### ISSUE TITLE
Additional verification of required fields in the bundle

### ISSUE THREAD

**Issue by Hayden-IO** [sigstore/sigstore-go#63]
**Description**

We currently check for an inclusion proof or promise in a [validation function](https://github.com/sigstore/sigstore-go/blob/f7518a68b37b7887b69e1607b9ddb668ddec117d/pkg/bundle/bundle.go#L74-L94) when constructing the bundle.

Further verification for required fields would be ideal. Using proto reflection, we could use the field annotation `field_behavior` to determine if the field has been annotated as required.

**Comment by vishal-chdhry** [comment]
@haydentherapper I was looking into this and I encountered an issue with `content` field. The protobuf bundle marks [all of the field in content as `REQUIRED`](https://github.com/sigstore/protobuf-specs/blob/32344d05040da6b71adb15806390329a179382b5/protos/sigstore_bundle.proto#L93-L97). Validation of content field fails for this [test case](https://github.com/sigstore/sigstore-go/pull/189#discussion_r1615260444).

The [proposal](https://google.aip.dev/203#:~:text=The%20use%20of%20REQUIRED%20indicates%20that%20the%20field%20must%20be%20present%20(and%20set%20to%20a%20non%2Dempty%20value)%20on%20the%20request%20or%20resource.) says that:

> The use of REQUIRED indicates that the field must be present (and set to a non-empty value) on the request or resource.

There is something wrong in either the implementation library or the bundle, but I am not sure, what do you think?

**Comment by kommendorkapten** [comment]
Try with this branch and see if that resolves the issue: https://github.com/sigstore/protobuf-specs/pull/337

**Comment by vishal-chdhry** [comment]
@kommendorkapten  It fixes that test but it makes an empty `VerificationMaterial` a valid payload
```golang
bundle: &ProtobufBundle{
	Bundle: &protobundle.Bundle{
		MediaType:            "application/vnd.dev.sigstore.bundle+json;version=0.3",
		VerificationMaterial: &protobundle.VerificationMaterial{},
		Content:              nil,
	},
},
```
```shell
$ go test -run=Test_BundleValidation .
--- FAIL: Test_BundleValidation (0.00s)
    --- FAIL: Test_BundleValidation/name:Empty_verification_material (0.00s)
        bundle_test.go:182: Protobuf.Bundle() error = <nil>, wantErr bundle validation failed: missing required field: verification_material.public_key
FAIL
FAIL    github.com/sigstore/sigstore-go/pkg/bundle      0.190s
FAIL
```
I think [this](https://github.com/sigstore/sigstore-go/pull/189/files#diff-adf66e7d61809c496270e551a0c2212876f514667c72d48a0d18cd08ce7dc6daR138-R147) test case should fail

**Comment by Hayden-IO** [comment]
I agree that an empty VerificationMaterials should cause a test failure as at least a key hint or certificate should be required. Could this be because we need to add some logic that checks for the one_of behavior extension value?

**Comment by vishal-chdhry** [comment]
> Could this be because we need to add some logic that checks for the one_of behavior extension value?

Yes, it might require some changes in the library I am using.
But since we have only 1 protobuf message to validate, and the message is simple, should I just write a validation function specific to that struct myself. This will cut a dependency to that library and it will be faster as it is specific to the struct and not general purpose.

@haydentherapper @kommendorkapten wdyt?

**Comment by kommendorkapten** [comment]
@vishal-chdhry I think that makes sense to extend the custom method we have, and avoid reflection. As you said, the number of messages is low, and the so is the general complexity. I'm in favour 👍

**Comment by steiza** [comment]
At the sigstore-go client meeting, we said this was a "should" (not must) for v1.0.

**Comment by Hayden-IO** [comment]
Will not be doing this for v1.0. Follow up is to look into [protovalidate](https://github.com/bufbuild/protovalidate/).
